vimode/src/cmds/fold.c

```diff
@@ -0,0 +1,130 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "cmds/fold.h"
+#include "utils.h"
+
+/* fold command does not depend on state of parents */
+#define FOLD_NONE		0
+/* fold command depend on state of parents */
+#define FOLD_PARENT		1
+/* fold command depend on state of contracted parent if it exists */
+#define FOLD_CONTRACTED		2
+
+static gint prepare_fold(CmdParams *p, gint filter)
+{
+	/* foldparent of the next line */
+	gint line = SSM(p->sci, SCI_GETFOLDPARENT, p->line + 1, 0);
+
+	if (p->line == line)
+		;  /* we are already on the fold point line */
+	else
+	{
+		/* foldparent of the current line */
+		line = SSM(p->sci, SCI_GETFOLDPARENT, p->line, 0);
+	}
+
+	/* retreive first parent when filter != FOLD_NONE
+	   when filter == FOLD_CONTRACTED we stop on first contracted parent if exist
+	*/
+	if (filter == FOLD_CONTRACTED && line != -1 && ! SSM(p->sci, SCI_GETFOLDEXPANDED, line, 0))
+		;	/* this fold point is contracted and filter == FOLD_CONTRACTED */
+	else if (filter != FOLD_NONE)
+	{
+		gint prev_line = line;
+		while (prev_line != -1)
+		{
+			prev_line = SSM(p->sci, SCI_GETFOLDPARENT, prev_line, 0);
+			if (prev_line != -1)
+			{
+				line = prev_line;
+				if (filter == FOLD_CONTRACTED && ! SSM(p->sci, SCI_GETFOLDEXPANDED, line, 0))
+					break;
+			}
+		}
+	}
+
+
+	if (line != -1)
+	{
+		/* move the cursor on the visible line before the fold */
+		gint pos = SSM(p->sci, SCI_POSITIONFROMLINE, line, 0);
+		SET_POS_NOX(p->sci, pos, TRUE);
+	}
+
+	return line;
+}
+
+
+void cmd_toggle_fold(CmdContext *c, CmdParams *p)
+{
+	gint line = prepare_fold(p, FOLD_NONE);
+	if (line != -1)
+		SSM(p->sci, SCI_FOLDLINE, (uptr_t) line, SC_FOLDACTION_TOGGLE);
+}
+
+
+void cmd_open_fold(CmdContext *c, CmdParams *p)
+{
+	gint line = prepare_fold(p, FOLD_NONE);
+	if (line != -1)
+		SSM(p->sci, SCI_FOLDLINE, (uptr_t) line, SC_FOLDACTION_EXPAND);
+}
+
+
+void cmd_close_fold(CmdContext *c, CmdParams *p)
+{
+	gint line = prepare_fold(p, FOLD_NONE);
+	if (line != -1)
+		SSM(p->sci, SCI_FOLDLINE, (uptr_t) line, SC_FOLDACTION_CONTRACT);
+}
+
+
+void cmd_toggle_fold_child(CmdContext *c, CmdParams *p)
+{
+	gint line = prepare_fold(p, FOLD_CONTRACTED);
+	if (line != -1)
+		SSM(p->sci, SCI_FOLDCHILDREN, (uptr_t) line, SC_FOLDACTION_TOGGLE);
+}
+
+
+void cmd_open_fold_child(CmdContext *c, CmdParams *p)
+{
+	gint line = prepare_fold(p, FOLD_NONE);
+	SSM(p->sci, SCI_FOLDCHILDREN, (uptr_t) line, SC_FOLDACTION_EXPAND);
+}
+
+
+void cmd_close_fold_child(CmdContext *c, CmdParams *p)
+{
+	gint line = prepare_fold(p, FOLD_PARENT);
+	if (line != -1)
+		SSM(p->sci, SCI_FOLDCHILDREN, (uptr_t) line, SC_FOLDACTION_CONTRACT);
+}
+
+
+void cmd_open_fold_all(CmdContext *c, CmdParams *p)
+{
+	SSM(p->sci, SCI_FOLDALL, SC_FOLDACTION_EXPAND | SC_FOLDACTION_CONTRACT_EVERY_LEVEL, 0);
+}
+
+
+void cmd_close_fold_all(CmdContext *c, CmdParams *p)
+{
```

Comment:
This will need `goto_above_fold(p, GOTO_TOPMOST_PARENT);` otherwise the fold where the cursor is won't get contracted because of what I mentioned before.